src/server/game/Entities/Unit/Unit.cpp

```diff
@@ -16179,3 +16186,33 @@ bool Unit::VisibleAuraSlotCompare::operator()(AuraApplication* left, AuraApplica
 {
     return left->GetSlot() < right->GetSlot();
 }
+
+bool Unit::IsNextSpellACrit(Unit* target, uint32 spellId) const
+{
+    auto itr = m_spellCritMap.find(std::pair<Unit*, uint32>(target, spellId));
+
+    return itr != m_spellCritMap.end() && itr->second == true;
+}
+
+void Unit::SetNextSpellCrit(Unit* target, uint32 spellId, bool crit)
+{
+    auto itr = m_spellCritMap.find(std::pair<Unit*, uint32>(target, spellId));
+
+    if (itr == m_spellCritMap.end())
+        m_spellCritMap.emplace(std::pair<Unit*, uint32>(target, spellId), crit);
+}
+
+bool Unit::HasNextSpellCritData(Unit* target, uint32 spellId) const
+{
+    auto itr = m_spellCritMap.find(std::pair<Unit*, uint32>(target, spellId));
+
+    return itr != m_spellCritMap.end();
+}
+
+void Unit::RemoveNextSpellCritData(Unit* target, uint32 spellId)
```

Comment:
what happens if the target creature is removed from world before the spell crit is used ? will m_spellCritMap still hold a reference to it ?